File: C#/insertionSort.cs
using System;

class InsertionSort
{
    // Method to perform Insertion Sort on an array
    public static void Sort(int[] array)
    {
        int n = array.Length; // Get the length of the array

        // Iterate through the array starting from the second element
        for (int i = 1; i < n; i++)
        {
            // Store the current element to be inserted
            int key = array[i];
            int j = i - 1; // Start comparing with the previous element

            // Move elements that are greater than the key to one position ahead
            while (j >= 0 && array[j] > key)
            {
                array[j + 1] = array[j]; // Shift element to the right
                j--; // Move to the previous element
            }

            // Insert the key at its correct position
            array[j + 1] = key;
        }
    }

    // Main method to test the Insertion Sort algorithm
    static void Main()
    {
        // Example array to be sorted
        int[] array = { 12, 11, 13, 5, 6 };

        Console.WriteLine("Original array:");
        PrintArray(array); // Print the original array

        // Call the sort method
        Sort(array); // Sort the array using Insertion Sort

        Console.WriteLine("Sorted array:");
        PrintArray(array); // Print the sorted array
    }

    // Method to print the elements of the array
    private static void PrintArray(int[] array)
    {
        foreach (int item in array)
        {
            Console.Write(item + " "); // Print each item followed by a space
        }
        Console.WriteLine(); // Print a new line after the array
    }
}

File: C++/insertionSort.cpp
#include <iostream>
#include <vector>

// Function to implement Insertion Sort
void insertionSort(std::vector<int>& arr) {
    int n = arr.size(); // Get the size of the array

    // Start from the second element (index 1), because the first element is trivially sorted
    for (int i = 1; i < n; ++i) {
        int key = arr[i]; // The current element to be inserted into the sorted part
        int j = i - 1;    // Start comparing with the element before the current one

        // Move elements of the sorted part that are greater than 'key' to one position ahead
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j]; // Shift the element to the right
            j--;                 // Move to the next element on the left
        }

        // Insert the current element ('key') at its correct position in the sorted part
        arr[j + 1] = key;
    }
}

// Helper function to print the elements of the array
void printArray(const std::vector<int>& arr) {
    for (int i = 0; i < arr.size(); ++i) {
        std::cout << arr[i] << " "; // Print each element followed by a space
    }
    std::cout << std::endl; // Print a newline at the end
}

// Main function
int main() {
    // Example array to sort
    std::vector<int> arr = {12, 11, 13, 5, 6};

    // Output the original array
    std::cout << "Original array: ";
    printArray(arr);

    // Call the insertion sort function
    insertionSort(arr);

    // Output the sorted array
    std::cout << "Sorted array: ";
    printArray(arr);

    return 0;
}
